scalanet/src/io/iohk/scalanet/peergroup/kademlia/KBuckets.scala

```diff
@@ -0,0 +1,109 @@
+package io.iohk.scalanet.peergroup.kademlia
+
+import java.util.concurrent.{ConcurrentHashMap, CopyOnWriteArraySet}
+
+import io.iohk.scalanet.peergroup.kademlia.KBuckets._
+import scodec.bits.BitVector
+
+import scala.collection.JavaConverters._
+import scala.collection.mutable
+
+class KBuckets(val baseId: BitVector) {
+
+  private val buckets = new ConcurrentHashMap[Int, KBucket]().asScala
+
+  add(baseId)
+
+  /**
+    * Find the n nodes closest to nodeId in kBuckets.
+    * Return the resulting node records sorted by distance from the nodeId.
+    * The indices into the kBuckets are defined by their distance from referenceNodeId.
+    */
+  def closestNodes(nodeId: BitVector, n: Int): List[BitVector] = {
+    // sketch
+    // the id range of a bucket is 2^i bits,
+    // where i is the common prefix length between the bucket prefix and the base nodeId.
+    // each kbucket covers a distance range [2^i, 2^i+1 - 1]
+    // this means that buckets with larger i cover larger ranges.
+    // so, for eg, bucket 1 covers 4-2-1=1 node whereas bucket 7 covers 256-128-1=127 nodes
+    // in general each bucket covers 2^i+1 - 2^i - 1 = 2^i - 1
+
+    // therefore when finding an arbitrary nodeId
+    // find the prefix
+
+    // replace with routing tree described in kademlia paper...
+    iterator.toList.sorted(new XorOrdering(nodeId)).take(n)
+  }
+
+  /**
+    * Add a node record into the KBuckets (TODO if there is capacity).
+    *
+    * @return this KBuckets instance.
+    */
+  def add(nodeId: BitVector): KBuckets = {
+    if (nodeId.length != this.baseId.length)
+      throw new IllegalArgumentException(
+        s"Illegal attempt to add node id with a length different than the this node id."
+      )
+
+    val d = Xor.d(nodeId, this.baseId)
+
+    val bucket: KBucket = buckets.getOrElseUpdate(d, newKBucket)
+
+    bucket.add(nodeId)
+
+    this
```

Comment:
I would prefer the method to return Unit.
The expression `bucket.add(nodeId)` has a side effect and the type signature of this method suggest that a new updated instance is returned without affecting the original object's state.

Reply:
I have copied the signatures from the scala collections. Agreed, they are not the most intuitive but I am keeping for consistency in case we want to extend a scala collection at some point.